server/conn.go

```diff
@@ -988,67 +988,77 @@ func (cc *clientConn) writeChunks(ctx context.Context, rs ResultSet, binary bool
 		if rowCount == 0 {
 			break
 		}
-		err = cc.writeRowsInChunk(chk, data, rs.Columns(), binary)
-		if err != nil {
-			return errors.Trace(err)
+		for i := 0; i < rowCount; i++ {
+			data = data[0:4]
+			if binary {
+				data, err = dumpBinaryRow(data, rs.Columns(), chk.GetRow(i))
+			} else {
+				data, err = dumpTextRow(data, rs.Columns(), chk.GetRow(i))
+			}
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if err = cc.writePacket(data); err != nil {
+				return errors.Trace(err)
+			}
 		}
 	}
 	return errors.Trace(cc.writeEOF(serverStatus))
 }
 
 // writeChunksWithFetchSize writes data from a Chunk, which filled data by a ResultSet, into a connection.
 // binary specifies the way to dump data. It throws any error while dumping data.
-// serverStatus, a flag bit represents server information
-// fetchSize, the desired number of rows to be fetched each time when client uses cursor
-func (cc *clientConn) writeChunksWithFetchSize(ctx context.Context, rs ResultSet, binary bool, serverStatus uint16, fetchSize int) error {
-	data := make([]byte, 4, 1024)
-	chk := rs.NewChunk()
-	fetchRows := 0
-	for {
+// serverStatus, a flag bit represents server information.
+// fetchSize, the desired number of rows to be fetched each time when client uses cursor.
+func (cc *clientConn) writeChunksWithFetchSize(ctx context.Context, rs ResultSet, serverStatus uint16, fetchSize int) error {
+	fetchedRows := rs.GetFetchedRows()
+	var curRows []chunk.Row
+
+	// if fetchedRows is empty, getting data from recordSet.
+	if len(fetchedRows) == 0 {
```

Comment:
As long as the amount of rows returned to the client is '<=' fetch size, and the flag `ServerStatusCursorExists`  is set,  the client can process the data correctly, and knows that the data has not yet been sent completely, it will continue to send the `COM_STMT_FETCH` request for the next batch of rows.

For this bulk streaming operation, a batch of rows smaller than fetch size is acceptable as long as the overall result is correct.

Reply:
You are right, let me try again

Reply:
Thanks. tommorrow I will do some test for this great job~ and try to push other guys to review this...thanks ^ ^